lib/fetcher.js

```diff
@@ -203,7 +208,27 @@ class FetcherBase {
     // gets to the point of re-setting the integrity.
     const istream = ssri.integrityStream(this.opts)
     istream.on('integrity', i => this.integrity = i)
-    return stream.on('error', er => istream.emit('error', er)).pipe(istream)
+    // we have to return a stream that gets ALL the data, and proxies errors,
+    // but then pipe from the original tarball stream into the cache as well.
+    // To do this without losing any data, and since the cacache put stream
+    // is not a passthrough, we have to pipe from the original stream into
+    // the cache AFTER we pipe into the istream.
+    // If the cacache put stream was a passthrough, we could just return
+    // a single pipeline with all three, but that would make cacache put
+    // streams less useful, because they'd have to be read from to work.
+    stream.on('error', er => istream.emit('error', er))
+    stream.pipe(istream)
+    if (this.opts.cache && this[_cacheFetches]) {
+      const cstream = cacache.put.stream(
+        this.opts.cache,
+        `pacote:tarball:${this.from}`,
+        this.opts
+      )
+      // best-effort.  cache write errors should not crash the fetch.
+      cstream.on('error', (er) => {})
+      stream.pipe(cstream)
```

Comment:
Also, not sure how streams work, but is `istream` inserted in between `stream` & `cstream` during the `stream.pipe(cstream)`. Maybe 

```suggestion
istream.pipe(cstream)
```

would be clearer if it does the same thing?

I'm working all of these off of my initial `stream.pipe(istream).pipe(cstream)`

Reply:
If we pipe the istream to the cstream, then the caller never gets the bytes, because they go into the cache writer and black-hole there.

We need the stream _itself_ to pipe to both so that the bytes go to both locations, but the user can read from the istream.